persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/interceptors/ProxyProtocolQueryInterceptor.java

```diff
@@ -254,14 +254,37 @@ private static ByteBuffer buildColumnValue(String name, InetAddress publicAddres
     }
   }
 
+  /**
+   * Get tokens generated for a public address.
+   *
+   * <p>This caches generated tokens for known peer addresses; otherwise, it re-calculates tokens
+   * every call for unknown addresses. This prevents {@code tokensCache} from growing without bound
+   * as {@code publicAddress} is provided by the client (proxy). This could use something like
+   * {@link LoadingCache} with maximum size, but then that requires exposing yet another
+   * configuration value to users.
+   *
+   * @param publicAddress
+   * @return a list of random token calculated using the the public address as a seed.
+   */
+  private Set<String> getTokens(InetAddress publicAddress) {
+    if (peers.contains(publicAddress)) {
+      return tokensCache.computeIfAbsent(
+          publicAddress,
+          pa -> StargateSystemKeyspace.generateRandomTokens(pa, DatabaseDescriptor.getNumTokens()));
+    } else {
```

Comment:
It's not entirely clear to me how we can hit this `else` branch. I.e. what "unknown" address means.

Following the caller hierarchy, the only code that leads to this method is `interceptQuery`, where the address is taken directly from the `peers` field.
```java
      Set<InetAddress> currentPeers = peers;
      ...
      for (InetAddress peer : currentPeers) {
        if (!peer.equals(publicAddress.getAddress())) {
          rows.add(buildRow(selectStatement.getResultMetadata(), peer));
        }
      }
```

Are we covering the case where `peers` was updated asynchronously before we could get to this method?

Reply:
`buildRow()` can happen on the `system.local` code path.

```
      rows =
          Collections.singletonList(
              buildRow(selectStatement.getResultMetadata(), publicAddress.getAddress())); // This comes from the client proxy
```